File: client/native/geolocation.js
// native/geolocation.js — Geolocation API glue for the N3 geolocation capability.

import { CapabilityError } from "./index.js";

/**
 * Map a GeolocationPositionError code to a stable capability error code.
 * @param {number} code
 * @returns {string}
 */
function geoErrorCode(code) {
  // 1 = PERMISSION_DENIED, 2 = POSITION_UNAVAILABLE, 3 = TIMEOUT
  if (code === 1) return "permission_denied";
  if (code === 3) return "timeout";
  return "unavailable";
}

/**
 * Request a single position fix.
 * @param {{high_accuracy:boolean}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{latitude:number,longitude:number,accuracy:number,altitude:number|null}>}
 * @throws {CapabilityError} permission_denied / timeout / unavailable / insecure_context.
 */
export async function geolocationGet(args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || !nav.geolocation) {
    throw new CapabilityError("unavailable", "geolocation is not available");
  }
  return new Promise((resolve, reject) => {
    nav.geolocation.getCurrentPosition(
      (pos) => {
        const c = pos.coords;
        resolve({
          latitude: c.latitude,
          longitude: c.longitude,
          accuracy: typeof c.accuracy === "number" ? c.accuracy : 0.0,
          altitude: typeof c.altitude === "number" ? c.altitude : null,
        });
      },
      (err) => reject(new CapabilityError(geoErrorCode(err && err.code), err && err.message)),
      { enableHighAccuracy: Boolean(args.high_accuracy) },
    );
  });
}

/**
 * Watch the device position, streaming a shaped payload per update (T-EV).
 *
 * Each fix emits `{ event: {latitude, longitude, accuracy, altitude} }`; a
 * geolocation error emits `{ error, message }` (permission denials map to
 * "permission_denied", everything else to "unavailable"). The returned function
 * clears the underlying `watchPosition` registration.
 *
 * @param {{high_accuracy?:boolean}} args
 * @param {(payload:Object) => void} emit  Sink for shaped stream payloads.
 * @param {import("./index.js").NativeDeps} deps
 * @returns {() => void}  Teardown that stops the position watch.
 * @throws {CapabilityError} unavailable — when the Geolocation watch API is absent.
 */
export function geolocationWatch(args, emit, deps) {
  const geo = (deps.navigator || /** @type {any} */ (globalThis).navigator).geolocation;
  if (!geo || !geo.watchPosition) {
    throw new CapabilityError("unavailable", "geolocation watch is not available");
  }
  const id = geo.watchPosition(
    (pos) =>
      emit({
        event: {
          latitude: pos.coords.latitude,
          longitude: pos.coords.longitude,
          accuracy: pos.coords.accuracy || 0,
          altitude: pos.coords.altitude ?? null,
        },
      }),
    (err) =>
      emit({
        error: err && err.code === 1 ? "permission_denied" : "unavailable",
        message: (err && err.message) || "",
      }),
    { enableHighAccuracy: !!args.high_accuracy },
  );
  return () => geo.clearWatch(id);
}

File: tempestweb/native/geolocation.py
"""Native geolocation capability over the browser's Geolocation Web API.

The web sibling of :mod:`tempestroid.native.geolocation`. :func:`get_position`
(aliased :func:`get`) sends a ``geolocation.get`` ``native_call``;
``client/native/geolocation.js`` calls ``navigator.geolocation.getCurrentPosition``
and replies with the fix.

Naming mirrors tempestroid (:class:`Position`, :func:`get_position`) so the same
application code reads identically on Android and on the web.
"""

from __future__ import annotations

from collections.abc import AsyncIterator

from pydantic import BaseModel, ConfigDict

from tempestweb.native.dispatch import native_events, send_native_call

__all__ = ["Position", "get", "get_position", "watch"]


class Position(BaseModel):
    """A geographic position fix returned by the browser.

    Mirrors ``GeolocationCoordinates``: ``accuracy`` is always present, while
    ``altitude`` is ``None`` when the device cannot report it.

    Attributes:
        latitude: Latitude in decimal degrees.
        longitude: Longitude in decimal degrees.
        accuracy: Horizontal accuracy radius in meters (``0.0`` if unknown).
        altitude: Altitude in meters above the WGS84 ellipsoid, or ``None``.
    """

    model_config = ConfigDict(frozen=True)

    latitude: float
    longitude: float
    accuracy: float = 0.0
    altitude: float | None = None


async def get_position(high_accuracy: bool = True) -> Position:
    """Request a single location fix from the browser.

    Args:
        high_accuracy: Set ``enableHighAccuracy`` (prefer GPS) when ``True``.

    Returns:
        The current :class:`Position`.

    Raises:
        NativeError: If the user denies permission (``permission_denied``), the
            page is not a secure context (``insecure_context``), or no fix is
            available (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("geolocation.get", {"high_accuracy": high_accuracy})
    return Position.model_validate(value)


async def watch(high_accuracy: bool = True) -> AsyncIterator[Position]:
    """Stream location fixes as the device moves (event channel / T-EV).

    Opens a ``navigator.geolocation.watchPosition`` subscription and yields a fresh
    :class:`Position` for every update until the ``async for`` loop is exited (which
    cancels the watch). Consume it with::

        async for pos in geolocation.watch():
            app.set_state(lambda s: setattr(s, "here", pos))

    Args:
        high_accuracy: Set ``enableHighAccuracy`` (prefer GPS) when ``True``.

    Yields:
        Each updated :class:`Position`.

    Raises:
        NativeError: If the user denies permission (``permission_denied``), the page
            is not a secure context (``insecure_context``), or the watch fails.
        BrowserUnavailableError: If no bridge is installed, or the installed bridge
            does not support the event channel.
    """
    args = {"high_accuracy": high_accuracy}
    async for value in native_events("geolocation.watch", args):
        yield Position.model_validate(value)


#: Alias matching the tempestroid/plan API ``await geolocation.get()``.
get = get_position
